### PR TITLE
SEQNG-543: Port to cats/cats-effect/fs2/http4s

### PR DESCRIPTION
**Pull request by cquiroz**
This is a fairly large PR containing a full port of the project using cats/cats-effect/fs2 and an updated http4s instead of the scalaz/scalaz-stream family. Note this is a one-to-one port, not much has been done to improve the code otherwise.

There are a few bits left as uncertain especially around fs2 interactions. Manual tests show all the pieces working but this needs a fair amount of testing with hardware before merging.

As much as possible law testing is done at the places it corresponds

There are still 2 scalaz implicit dependencies (spmodel libraries, and scalajs-react-extras) but scalaz is out of the classpath. The scalajs-react dependency will be removed shortly when we upgrade to scalajs-react 1.2 with react 16.

### CODE REVIEW

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala [hunk 1]
```diff
@@ -31,55 +34,47 @@ class Engine[D: ActionMetadataGenerator, U](implicit ev: ActionMetadataGenerator
...
+   * Streams as output.
+   *
    * Its type parameters are:
    * A: Type of the output (usually Unit)
    * D: Type of the user data included in the state machine state.
    *
    * Making it final causes an error: "The outer reference in this type test cannot be checked at run time"
    */
-  case class HandleP[A](run: Handle[(A, Option[Process[Task, EventType]])])
+  case class HandleP[A](run: Handle[(A, Option[Stream[IO, EventType]])])
```
Comment (tpolecat) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala]: Just a general note, when using tuples you should always put any free type variable in the right most position; i.e., `(Foo, A)` rather than `(A, Foo)` because the available instances assume right biasing.

Comment (cquiroz) [reply modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala]: what are the available instances?

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala [hunk 2]
```diff
@@ -31,55 +34,47 @@ class Engine[D: ActionMetadataGenerator, U](implicit ev: ActionMetadataGenerator
...
-      fa.run.flatMap{
+    override def flatMap[A, B](fa: HandleP[A])(f: A => HandleP[B]): HandleP[B] = HandleP[B](
+      fa.run.flatMap {
         case (a, op1) => f(a).run.map{
           case (b, op2) => (b, concatOpP(op1, op2))
         }
       }
     )
+
+    override def tailRecM[A, B](a: A)(f: A => HandleP[Either[A, B]]): HandleP[B] = ??? //scalastyle:ignore
```
Comment (tpolecat) [inline modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala]: I think this is right.

```scala
def tailRecM[A, B](a: A)(f: A => HandleP[Either[A, B]]): HandleP[B] = {
  
  // We don't really care what this type is
  type Unused = Option[Stream[IO, EventType]]
  
  // Construct a StateT that delegates to IO's tailRecM
  val st: StateT[IO, StateType, (B, Unused)] =
    StateT { s =>
      Monad[IO].tailRecM[(StateType, A), (StateType, (B, Unused))]((s, a)) { 
        case (s, a) => 
          f(a).run.run(s).map {
            case (sʹ, (Left(a), u))  => Left((sʹ, a))
            case (sʹ, (Right(b), u)) => Right((sʹ, (b, u)))
          }
      }
    }
  
  // Done
  HandleP(st)
  
}
```

**Comment by tpolecat** [comment]
The build expects `sbt-updates` but it looks like it didn't make it into `plugins.sbt`

File: modules/edu.gemini.web.server.common/src/test/scala/edu/gemini/web/common/http4s/StaticRoutesSpec.scala [hunk 3]
```diff
@@ -47,47 +48,47 @@ class StaticRoutesSpec extends FlatSpec with Matchers with EitherValues with Uri
...
     }
     it should "return a generated index(true) with development files on dev mode" in {
       val service = new StaticRoutes(index(true), true, builtAtMillis).service
-      service.apply(Request(uri = uri("/"))).unsafePerformSync.orNotFound.status should equal (Status.Ok)
-      service.apply(Request(uri = uri("/"))).unsafePerformSync.orNotFound.headers should contain (`Content-Type`(`text/html`, Charset.`UTF-8`))
-      service.apply(Request(uri = uri("/"))).unsafePerformSync.orNotFound.headers should contain (`Cache-Control`(NonEmptyList(`no-cache`())))
-      val body = service.apply(Request(uri = uri("/"))).unsafePerformSync.orNotFound.body.runLast.unsafePerformSync.map(_.decodeUtf8).fold("")(_.right.value)
-      body should include regex ".*jsdeps\\.js.*"
-      body should include regex ".*app*\\.js.*"
+      service.apply(Request(uri = uri("/"))).value.map(_.map(_.status)).unsafeRunSync should contain(Status.Ok)
```
Comment (tpolecat) [inline modules/edu.gemini.web.server.common/src/test/scala/edu/gemini/web/common/http4s/StaticRoutesSpec.scala]: FYI `fa.map(_.map(f))` can be written `Nested(fa).map(f).value` which is sometimes nicer. What you have here is fine though.

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala [hunk 4]
```diff
@@ -38,20 +38,25 @@ package engine {
...
+    case object Started extends ActionState
     final case class Paused[C <: PauseContext](ctx: C) extends ActionState
-
     final case class Completed[V <: RetVal](r: V) extends ActionState
-
     final case class Failed(e: Error) extends ActionState
 
+    object ActionState {
+      // TODO: This is not always valid
+      implicit val equal: Eq[ActionState] = Eq.fromUniversalEquals
```
[the remark this answers is not in the document]
Comment (tpolecat) [reply modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala]: Could also have a `.isIdle` method, which might be nicer.

File: modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.client/src/main/scala/edu/gemini/seqexec/web/client/components/sequence/toolbars/SequenceDefaultToolbars.scala [hunk 5]
```diff
@@ -80,8 +79,8 @@ object SequenceControl {
         control.whenDefined { m =>
           val ControlModel(id, isPartiallyExecuted, nextStep, status, inConflict) = m
           val nextStepToRun = nextStep.getOrElse(0) + 1
-          val runContinueTooltip = s"${isPartiallyExecuted ? "Continue" | "Run"} the sequence from the step $nextStepToRun"
-          val runContinueButton = s"${isPartiallyExecuted ? "Continue" | "Run"} from step $nextStepToRun"
+          val runContinueTooltip = s"${isPartiallyExecuted.fold("Continue", "Run")} the sequence from the step $nextStepToRun"
+          val runContinueButton = s"${isPartiallyExecuted.fold("Continue", "Run")} from step $nextStepToRun"
```
[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.client/src/main/scala/edu/gemini/seqexec/web/client/components/sequence/toolbars/SequenceDefaultToolbars.scala]: I kind of like `fold` here but `Option.fold` has the order inverted which is very confusing

File: modules/edu.gemini.web.server.common/src/test/scala/edu/gemini/web/common/http4s/StaticRoutesSpec.scala (hunk 3 above)
Comment (cquiroz) [reply modules/edu.gemini.web.server.common/src/test/scala/edu/gemini/web/common/http4s/StaticRoutesSpec.scala]: Will `Nested` work on more than two levels?

**Comment by cquiroz** [comment]
The removal of scalaz takes 275k of the unoptimized js. This will grow even further once we get rid of the scalaz dependency through `scalajs-react`

**Comment by cquiroz** [comment]
`sbt-updates` was added on the previous PR

File: modules/edu.gemini.web.server.common/src/test/scala/edu/gemini/web/common/http4s/StaticRoutesSpec.scala (hunk 3 above)
Comment (tpolecat) [reply modules/edu.gemini.web.server.common/src/test/scala/edu/gemini/web/common/http4s/StaticRoutesSpec.scala]: No, it just goes down one level.

```scala
@ Nested(Nested(Option(Option(Option(1))))).map(_ + 1).value.value 
res13: Option[Option[Option[Int]]] = Some(Some(Some(2)))
```

File: modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala (hunk 1 above)
Comment (cquiroz) [reply modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala]: Do you think we should flip this to be 
```scala
case class HandleP[A](run: Handle[(Option[Stream[IO, EventType]], A)])
```

Comment (tpolecat) [reply modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala]: I wouldn't worry about it right now. It was just an observation.

Comment (cquiroz) [reply modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala]: Is it also true for higher arities? `(A, B, C, ?)`

Comment (tpolecat) [reply modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala]: In principle yes, but Cats doesn't define those instances. `(A, B, C, ?)` is isomorphic to `((A, B, C), ?)` so you can always coerce it into that shape temporarily if you need to.

File: modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala [hunk 6]
```diff
@@ -20,7 +23,48 @@ object events {
   }
 
   object SeqexecModelUpdate {
-    implicit val equal: Equal[SeqexecModelUpdate] = Equal.equalA
+    import SeqexecEvent._
+    implicit val equal: Eq[SeqexecModelUpdate] =
+      Eq[Either[SequenceStart,
```
[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: This is weird. changing this to `Eq.by` produces a:
```
[error] Caused by: scala.UninitializedFieldError: Uninitialized field: /Users/cquiroz/Projects/ocs3/modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala: 27
```

When I run the tests

[the remark this answers is not in the document]
Comment (tpolecat) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: Curious, that means it's calling itself recursively.

[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: Yes. I'm pulling my hair to find out the reason, but I think I'm going to give up and do `Eq.instance`

**Comment by tpolecat** [comment]
No, it's `sbt-updates`. This is weird. If you do `dependencyUpdates` at the sbt prompt does it work? Your branch's build doesn't even load for me unless I pass `-Docs3.skipDependencyUpdates`.

**Comment by cquiroz** [comment]
now i know, I have it locally on my `.sbt/` dir. I'll send an update to fix this. Thanks for noticing it

File: modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: Yes they do run. My new instance is this one:
```
    implicit val equal: Eq[SeqexecEvent] =
      Eq.by {
        case s: ConnectionOpenEvent    => s
        case s: SequenceStart          => s
        case s: StepExecuted           => s
        case s: FileIdStepExecuted     => s
        case s: SequenceCompleted      => s
        case s: SequenceLoaded         => s
        case s: SequenceUnloaded       => s
        case s: StepBreakpointChanged  => s
        case s: OperatorUpdated        => s
        case s: ObserverUpdated        => s
        case s: ConditionsUpdated      => s
        case s: StepSkipMarkChanged    => s
        case s: SequencePauseRequested => s
        case s: SequencePauseCanceled  => s
        case s: SequenceRefreshed      => s
        case s: ActionStopRequested    => s
        case s: ResourcesBusy          => s
        case s: SequenceUpdated        => s
        case s: SequencePaused         => s
        case s: ExposurePaused         => s
        case s: SequenceError          => s
        case s: NewLogMessage          => s
        case s: ServerLogMessage       => s
        case s: NullEvent.type         => s
      }
```

[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: That one produces an `UnitializedError`. I'm perhaps just missing a detail, but i suspect in this case `by` is pulling a bunch of `Eq` instances per each subclass and one or more of them are not yet initialized

[the remark this answers is not in the document]
Comment (tpolecat) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: Oh, the LUB of all those subtypes is `SeqexecEvent` so it's looking itself up recursively.

[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: What's `LUB`?

[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: Any ideas on how to workaround this? `Eq.instance` perhaps?

[the remark this answers is not in the document]
Comment (tpolecat) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: I think `.by` nested `Either` is the right way to do it, or introduce [kittens](https://github.com/milessabin/kittens) and the derivation can be done automatically. It might make sense to start using kittens for these mechanical instances.

```scala
implicit val equal: Eq[SeqexecModelUpdate] =
  cats.derived.semi.eq
```

[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: `kittens` sounds good in this context

[the remark this answers is not in the document]
Comment (tpolecat) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: I just tried it. Seems to work.

[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: perhaps I can do `Eq.instance` and rely on the `Eq` instances of subclasses members. I'm a bit wary of using macros as in kittens. but if that doesn't work autoderivation could be used

[the remark this answers is not in the document]
Comment (tpolecat) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: Doing the cases pairwise with `Eq.instance` will work but you will need a catch-all case that will defeat exhaustiveness checking. But as long as you have a good `Arbitrary` and check the `Eq` laws I think this is fine.

[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala]: nice `Eq.instance` doesn't break and all test are passing

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqexecEngine.scala [hunk 7]
```diff
@@ -74,81 +71,74 @@ class SeqexecEngine(settings: SeqexecEngine.Settings) {
...
+      val fd = Duration(settings.odbQueuePollingInterval.toSeconds, TimeUnit.SECONDS)
+      scheduler.fixedRate[IO](fd).flatMap { _ =>
+        Stream.emit(Event.getState[executeEngine.ConcreteTypes](refreshSequenceList()))
+      }
+    }
 
-  def eventProcess(q: EventQueue): Process[Task, SeqexecEvent] =
-    executeEngine.process(wye(q.dequeue, seqQueueRefreshProcess)(mergeHaltBoth))(Engine.State.empty[EngineMetadata](EngineMetadata.default)).flatMap(x => Process.eval(notifyODB(x))).map {
+  def eventStream(q: EventQueue): Stream[IO, SeqexecEvent] = {
+    executeEngine.process(q.dequeue.merge(seqQueueRefreshStream))(Engine.State.empty[EngineMetadata](EngineMetadata.default)).flatMap(x => Stream.eval(notifyODB(x))).map {
```
Comment (jluhrs) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqexecEngine.scala]: `mergeHaltBoth` was used to be able to stop the Stream by closing `q`. That is not possible if `merge` is used, because `seqQueueRefreshStream` is an infinite Stream and the Stream produced by `merge` only stops when both `Stream`s stop.

Comment (cquiroz) [reply modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqexecEngine.scala]: Feel free to change it

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gcal/GcalController.scala [hunk 8]
```diff
@@ -23,50 +21,59 @@ trait GcalController {
...
     case object Off extends LampState
 
     case object On extends LampState
 
+    implicit val eq: Eq[LampState] =
+      Eq.fromUniversalEquals
+
   }
 
-  final case class ArLampState(self: LampState) extends AnyVal
```
Comment (jluhrs) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gcal/GcalController.scala]: Why did you remove the `extends AnyVal` ?

Comment (cquiroz) [reply modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gcal/GcalController.scala]: It doesn't benefit you too much with a case class

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqexecEngine.scala [hunk 9]
```diff
@@ -74,81 +71,74 @@ class SeqexecEngine(settings: SeqexecEngine.Settings) {
 
   private val translator = SeqTranslate(settings.site, systems, translatorSettings)
 
-  def load(q: EventQueue, seqId: SPObservationID): Task[SeqexecFailure \/ Unit] = loadEvents(seqId).flatMapF(q.enqueueAll(_).map(_.right)).run
+  def load(q: EventQueue, seqId: SPObservationID): IO[Either[SeqexecFailure, Unit]] =
+    loadEvents(seqId).flatMapF(b => q.enqueue(Stream.emits(b)).map(_.asRight).compile.last.attempt.map(_.bimap(SeqexecFailure.SeqexecException.apply, _ => ()))).value
```
[the remark this answers is not in the document]
Comment (cquiroz) [reply modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqexecEngine.scala]: Maybe that's the reason why fatal exceptions aren't being processed properly. Would you like to play with this part to improve it?

**Comment by cquiroz** [comment]
Great, we can merge this. We'll address the issues uncovered on separate PRs
